MelonLoader/Core.cs

```diff
@@ -21,6 +22,70 @@ internal static class Core
+
+        // First step of the Mono Core.Start entrypoint: Harmony patch the main Unity assembly with a suitable hook
+        private static void OnAssemblyLoad(object sender, AssemblyLoadEventArgs args)
+        {
+            const string sceneManagerTypeName = "UnityEngine.SceneManagement.SceneManager";
+            const string displayTypeName = "UnityEngine.Display";
+            
+            var assembly = args.LoadedAssembly;
+            var assemblyName = assembly.GetName().Name;
+
+            if (assemblyName is not ("UnityEngine.CoreModule" or "UnityEngine"))
+                return;
+
+            try 
+            { 
+                AppDomain.CurrentDomain.AssemblyLoad -= OnAssemblyLoad;
+
+                var sceneManagerType = assembly.GetType(sceneManagerTypeName, false);
+                if (sceneManagerType != null)
+                {
+                    MonoCoreStartHookMethod = sceneManagerType.GetMethod("Internal_ActiveSceneChanged", BindingFlags.NonPublic | BindingFlags.Static);
+                    HarmonyInstance.Patch(MonoCoreStartHookMethod, prefix: new HarmonyMethod(typeof(Core), nameof(Entrypoint)));
+                    MelonLogger.Msg($"Hooked into {MonoCoreStartHookMethod.FullDescription()}");
+                    return;
+                }
+
+                var displayType = assembly.GetType(displayTypeName, false);
+                if (displayType != null)
+                {
+                    MonoCoreStartHookMethod = displayType.GetMethod("RecreateDisplayList", BindingFlags.NonPublic | BindingFlags.Static);
+                    HarmonyInstance.Patch(MonoCoreStartHookMethod, postfix: new HarmonyMethod(typeof(Core), nameof(Entrypoint)));
+                    MelonLogger.Msg($"Hooked into {MonoCoreStartHookMethod.FullDescription()}");
+                    return;
+                }
+
+                MelonLogger.Error($"Couldn't find a suitable Core.Start entrypoint in the {assemblyName} assembly because " +
+                                  $"{sceneManagerTypeName} or {displayTypeName} do not exist in the assembly");
+            }
+            catch (Exception e)
+            {
+                MelonLogger.Error($"Unexpected error occured when trying to hook into {assemblyName}: {e}");
+            }
+        }
+
+        // Second step of the Mono Core.Start entrypoint: undo the Harmony patch and call the Core.Start method
+        private static void Entrypoint()
+        {
+            if (MonoCoreStartEntrypointAlreadyCalled)
+                return;
+
+            MonoCoreStartEntrypointAlreadyCalled = true;
+            try
+            {
+                HarmonyInstance.Unpatch(MonoCoreStartHookMethod, HarmonyPatchType.All, BuildInfo.Name);
+            }
+            catch (Exception e)
+            {
+                MelonLogger.Error($"Unexpected error when trying to unhook the Core.Start entrypoint: {e}");
+            }
+
+            Start();
+        }
```

Comment:
Make a new class specifically meant for Mono initialization and move this there.
This way the project is more organized and runtime-specific operations aren't done in the Core class.
Only compile the class if built for Mono (so under .net 6).

Also, move the event hook to a new Init method in that class and call it from this Core Init method